prow/pod-utils/gcs/options.go

```diff
@@ -0,0 +1,197 @@
+}
+
+// Options exposes the configuration necessary
+// for defining where in GCS an upload will land.
+type Options struct {
+	// Items are files or directories to upload
+	Items []string
+
+	// SubDir is appended to the resolved GCS path
+	SubDir string
+
+	// PathStrategy and the default Org and Repo
+	// determine how the GCS path is resolved from
+	// Prow-specified environment variables
+	PathStrategy string
+	DefaultOrg   string
+	DefaultRepo  string
+
+	GcsBucket          string
+	GceCredentialsFile string
+	DryRun             bool
+}
+
+func (o *Options) Complete(args []string) {
+	o.Items = args
+}
+
+func (o *Options) Validate() error {
+	if !o.DryRun {
+		if o.GcsBucket == "" {
+			return errors.New("GCS upload was requested but required flag --gcs-bucket was unset")
+		}
+
+		if o.GceCredentialsFile == "" {
+			return errors.New("GCS upload was requested but required flag --gcs-credentials-file was unset")
+		}
+	}
+
+	strategy := pathStrategyType(o.PathStrategy)
+	if strategy != pathStrategyLegacy && strategy != pathStrategyExplicit && strategy != pathStrategySingle {
+		return fmt.Errorf("GCS path strategy must be one of %q, %q, or %q", pathStrategyLegacy, pathStrategyExplicit, pathStrategySingle)
+	}
+
+	if strategy != pathStrategyExplicit && (o.DefaultOrg == "" || o.DefaultRepo == "") {
+		return fmt.Errorf("default org and repo must be provided for GCS strategy %q", strategy)
+	}
+
+	return nil
+}
+
+// Run will upload files to GCS as prescribed by
+// the options. Any extra files can be passed as
+// a parameter and will have the prefix prepended
+// to their destination in GCS, so the caller can
+// operate relative to the base of the GCS dir.
+func (o *Options) Run(extra map[string]UploadFunc) error {
+	var builder RepoPathBuilder
+	switch pathStrategyType(o.PathStrategy) {
+	case pathStrategyExplicit:
+		builder = NewExplicitRepoPathBuilder()
+	case pathStrategyLegacy:
+		builder = NewLegacyRepoPathBuilder(o.DefaultOrg, o.DefaultRepo)
+	case pathStrategySingle:
+		builder = NewSingleDefaultRepoPathBuilder(o.DefaultOrg, o.DefaultRepo)
+	}
+
+	spec, err := pjutil.ResolveSpecFromEnv()
+	if err != nil {
+		return fmt.Errorf("could not resolve job spec: %v", err)
+	}
+
+	var gcsPath string
+	jobBasePath := PathForSpec(spec, builder)
+	if o.SubDir == "" {
+		gcsPath = jobBasePath
+	} else {
+		gcsPath = path.Join(jobBasePath, o.SubDir)
+	}
+
+	uploadTargets := map[string]UploadFunc{}
+
+	// ensure that an alias exists for any
+	// job we're uploading artifacts for
+	if alias := AliasForSpec(spec); alias != "" {
+		uploadTargets[alias] = DataUpload(strings.NewReader(jobBasePath))
+	}
+
+	for _, item := range o.Items {
+		info, err := os.Stat(item)
+		if err != nil {
+			logrus.Warnf("Encountered error in resolving items to upload for %s: %v", item, err)
+			continue
+		}
+		if info.IsDir() {
+			gatherArtifacts(item, gcsPath, uploadTargets)
+		} else {
+			uploadTargets[path.Join(gcsPath, info.Name())] = FileUpload(item)
```

Comment:
This is only for items in the base directory. Any items underneath a directory specified to the tool retain their directory prefixes as relative paths from the base directory specified to the tool. In the majority of cases in Origin we are using this right now as a stop-gap to upload an `artifacts/` directory, so that just populates paths in GCS under `<test-base>/artifacts/`. I agree if you pointed the tool at multiple (regular) files in different places that all had the same name they'd collide in GCS but I'm not sure there is a use-case for that. We could make the mapping from file path to relative path under the GCS job prefix more explicit in the input to this tool, if you think that is useful.

Reply:
We could also calculate that a collision would occur and error out